File: epi_judge_python/remove_duplicates_from_sorted_list.py
from typing import Optional

from list_node import ListNode
from test_framework import generic_test


def remove_duplicates(L: ListNode) -> Optional[ListNode]:
    it1 = L
    if L and L.next:
        it2 = L.next
        while it2:
            if it1.data == it2.data:
                it1.next = it2.next
            else:
                it1 = it2
            it2 = it2.next

    return L


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main(
            'remove_duplicates_from_sorted_list.py',
            'remove_duplicates_from_sorted_list.tsv', remove_duplicates))

File: epi_judge_cpp/remove_duplicates_from_sorted_list.cc
#include <memory>

#include "list_node.h"
#include "test_framework/generic_test.h"
using std::shared_ptr;

shared_ptr<ListNode<int>> RemoveDuplicates(const shared_ptr<ListNode<int>> &L) {
  if (L && L->next) {
    auto fr{L->next}, ba{L};
    while (fr) {
      if (fr->data == ba->data) {
        ba->next = fr->next;
        auto t = fr;
        fr = fr->next;
        t.reset();
      } else {
        fr = fr->next;
        ba = ba->next;
      }
    }
  }
  return L;
}

int main(int argc, char *argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"L"};
  return GenericTestMain(args, "remove_duplicates_from_sorted_list.cc",
                         "remove_duplicates_from_sorted_list.tsv",
                         &RemoveDuplicates, DefaultComparator{}, param_names);
}
